### PR TITLE
@ - expressions in class method arguments

### PR DESCRIPTION
**Pull request by tsuchikazu**
`@` arguments doesn't work.

coffeescript
```coffeescript
class A
  constructor: (@b) ->
```

before es6 output
```javascript
class A {
  constructor(this.b) {}
}
```

after es6 output
```javascript
class A {
  constructor(b) {
    this.b = b;
  }
}
```

### CODE REVIEW

**Comment by juliankrispel** [comment]
Thanks. I'll have a look shortly.

File: src/parser.js [hunk 1]
```diff
@@ -465,20 +465,19 @@ function mapInArrayExpression(node, meta) {
   );
 }
 
-function extractArgumentMemberAssignment(nodes) {
+function extractAssignStatementsByArguments(nodes) {
```
Comment (juliankrispel) [inline src/parser.js]: not sure about the function name why does it make more sense?

**Comment by juliankrispel** [comment]
@tsuchikazu thanks for this

So actually there's a bunch of behaviour missing here not just @. right now `mapClassBodyElement` on line 225 is creating class methods, but actually it should be using the logic inside `mapFunction` on line 699, which deals with `@` variables in argument bodies and splats and expansions and a whole lot of other things unique to coffeescript argument bodies.

I'd rather use stuff that's already there for this.

Am I making sense?

**Comment by tsuchikazu** [comment]
@juliankrispel Thank you for your review!

I'm sorry for confusing you with my bad explanation. This PR scope is not only @ expressions in class method arguments. This PR scope is @ expressions in function arguments which doesn't have expansions and splats.

When function arguments has expansions or splats, it is used `mapArgumentsWithExpansion` logic. 
When function arguments doesn't have expansions or splats, it is used `extractAssignStatementsByArguments` logic of this PR.

Now master branch, arguments assign to this, when arguments has @ expression and default value. 
like this

coffee
```coffee
fn = (@a = 123) => 
```
es6
```javascript
var fn = (a = 123) => {
  this.a = a;
};
```

If merge this PR, arguments assign to this, when arguments has @ expression only.
coffee
```coffee
fn = (@a) => 
```
es6
```javascript
var fn = (a) => {
  this.a = a;
};
```

**Comment by juliankrispel** [comment]
ok sry I'll have another closer look and get back to you. I'm working now so expect to hear from me tomorrow

**Comment by tsuchikazu** [comment]
I added test for this PR at `FunctionExpression`. Please see also added commit.

**Comment by juliankrispel** [comment]
there are one or two bits missing here but I'll merge this and add them in a sec as a separate pr

**Comment by tsuchikazu** [comment]
Thank you!! :smile:

**Comment by juliankrispel** [comment]
@tsuchikazu WHY THANK YOU! Release coming shortly

**Pull request merged**